docs/modules/ROOT/pages/references/projectsyn/developer.adoc

```diff
@@ -0,0 +1,136 @@
+
+[NOTE]
+====
+When using the `commodore component new` command to bootstrap a new component, the `component-` prefix will be added automatically and _must not_ be specified:
+
+[source,console]
+----
+commodore component new openshift4-registry --name "OpenShift 4 Registry" --owner appuio
+----
+====
+
+
+== https://docs.openshift.com/container-platform/latest/operators/understanding_olm/olm-understanding-olm.html[Operator Lifecycle Manager (OLM)]
+
+We _recommend_ to avoid using the OLM to install operators on a cluster.
+This is mainly due to the added complexity the OLM introduces and us not using any of it's features. This might change in the future.
+
+Exceptions to this rule include cases where the Red Hat documentation explicitly requires the OLM.
+
+Any operators we **do** install with the OLM _should_ be restricted to single namespaces and _must not_ be installed using the `AllNamespaces` InstallMode.
+
+We don't provide any support for operators installed by customers via the OLM.
+
+
+== CustomResource (CR) Specs
+
+Components which configure one or multiple CRs _should_ do so by exposing its `.spec` field as a parameter.
+This enables the configuration of all possible fields via the inventory and provides the maximum flexibility.
+Default values _may_ either be set in the `defaults.yml` file of the component or in Jsonnet code.
+It _should_ always be possible to override default values in the inventory.
+
+[CAUTION]
+====
+Lists in the inventory get extended by http://reclass.pantsfullofunix.net/operations.html#data-merging[reclass] when being merged together.
+This means that the items of each list are added together in a new list.
+It can lead to undesired behavior if not accounted for.
+====
+
+If creating multiple CRs, a dict _should_ be used where its keys map to the names used for the CRs.
+Iterating over the keys of this dict in Jsonnet enables the enrichment of possible default values if not set.
+
+Due to the way http://reclass.pantsfullofunix.net/index.html[reclass] handels `null` overrides, the component _must_ handle `null` values in dicts and lists gracefully.
+
+.main.jsonnet
+[source,jsonnet]
+----
+local ingressControllers =
+  if params.ingressControllers != null
+  then std.objectFields(params.ingressControllers)
+  else [];
+
+{
+  [name]:
+    [kube._Object('operator.openshift.io/v1', 'IngressController', name) {
+      metadata+: {
+        namespace: params.namespace,
+      },
+      spec: {
+        defaultCertificate: {
+          name: acmeCertName,
+        },
+      } + params.ingressControllers[name],
+    }]
+  for name in ingressControllers
+} + {
+  [if std.length(ingressControllers) == 0 then '.gitkeep']: {},
+}
+----
+
+
+== Inventory Parameters
+
+A component _may_ define various parameters in the inventory (`defaults.yml`).
+These _should_ be namespaced in a dict with the component's name (hyphens replaced with underscores):
```

Comment:
```suggestion
These _must_ be namespaced in a dict with the component's name (hyphens replaced with underscores):
```

Reply:
I thought about this too but decided to stick with `should` because it might make sense to define other dicts as well. Especially for components which are tightly coupled like synsights or crossplane/crossplane-cloudscale.

Reply:
Both Synsights and Crossplane are early adopter components. Since then we have learned and I question what they are doing.

Mabybe we could go with _must_ but relax it by the following.

> A group of components _may_ agree upon a common prefix to share parameters between each other.